R/fit.R

```diff
@@ -919,3 +918,102 @@ CmdStanVB <- R6::R6Class(
+
+#' CmdStanGQ objects
+#'
+#' @name CmdStanGQ
+#' @family fitted model objects
+#' @template seealso-docs
+#'
+#' @description A `CmdStanGQ` object is the fitted model object returned by the
+#'   [`$generate_quantities()`][model-method-generate-quantities] method of a
+#'   [`CmdStanModel`] object.
+#'
+#' @details
+#' `CmdStanGQ` objects have the following methods:
+#'
+#' \tabular{ll}{
+#'  **Method** \tab **Description** \cr
+#'  [`$draws()`][fit-method-draws] \tab Returns the generated quantities
+#'  as a [`draws_matrix`][posterior::draws_matrix]. \cr
+#'  [`$summary()`][fit-method-summary] \tab
+#'  Run [`posterior::summarise_draws()`][posterior::draws_summary]. \cr
+#'  [`$save_object()`][fit-method-save_object]
+#'    \tab Save fitted model object to a file. \cr
+#'  [`$save_output_files()`][fit-method-save_output_files]
+#'    \tab Save output CSV files to a specified location. \cr
+#'  [`$save_data_file()`][fit-method-save_data_file]
+#'    \tab Save JSON data file to a specified location. \cr
+#'  [`$time()`][fit-method-time] \tab Report the total run time. \cr
+#'  `$output()` \tab Pretty print the output that was printed during generating quantities. \cr
+#' }
+#'
+CmdStanGQ <- R6::R6Class(
+  classname = "CmdStanGQ",
+  inherit = CmdStanFit,
+  public = list(
+    fitted_params_files = function() {
+      self$runset$args$method_args$fitted_params
+    },
+    num_chains = function() {
+      super$num_procs()
+    },
+    draws = function(variables = NULL) {
+      if (!length(self$output_files(include_failed = FALSE))) {
+        stop("No chains finished successfully. Unable to retrieve the generated quantities.", call. = FALSE)
+      }
+      to_read <- remaining_columns_to_read(
+        requested = variables,
+        currently_read = dimnames(private$draws_)$variable,
+        all = private$metadata_$model_params
+      )
+      if (is.null(to_read) || any(nzchar(to_read))) {
+        private$read_csv_(variables = to_read)
+      }
+      if (is.null(variables)) {
+        variables <- private$metadata_$model_params
+      }
+      matching_res <- matching_variables(variables, private$metadata_$model_params)
+      if (length(matching_res$not_found)) {
+        stop("Can't find the following variable(s) in the output: ",
+             paste(matching_res$not_found, collapse = ", "), call. = FALSE)
+      }
+      variables <- repair_variable_names(matching_res$matching)
+      private$draws_[,,variables]
+    },
+    output = function(id = NULL) {
+      if (is.null(id)) {
+        self$runset$procs$proc_output()
+      } else {
+        cat(paste(self$runset$procs$proc_output(id), collapse="\n"))
+      }
+    }
+  ),
+  private = list(
+    # inherits draws_ and metadata_ slots from CmdStanFit
+    read_csv_ = function(variables = NULL) {
+      variables_to_read <-
+      remaining_columns_to_read(
+        variables,
+        dimnames(private$draws_)$variable,
+        private$metadata_$model_params
+      )
```

Comment:
Do we need to use `remaining_columns_to_read()` again here when `read_csv_()` is being called by `draws()`, which already calls `remaining_columns_to_read()`? I recall that maybe we do, but it's not immediately obvious to me here why we do.

Reply:
Hm yeah, this is mostly that other functions that calll read_csv_() only read from CSV if necessary. We could get rid of the call in draws() though.